File: Android/app/src/main/java/com/max/spirihin/mytracksdb/Track.java
package com.max.spirihin.mytracksdb;

import android.location.Location;
import android.util.JsonReader;
import android.util.JsonWriter;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

import java.io.File;
import java.io.IOException;
import java.io.StringReader;
import java.io.StringWriter;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

public class Track {

    private ArrayList<TrackPoint> mPoints;

    public Track() {
        mPoints = new ArrayList<TrackPoint>();
    }

    public TrackPoint addPoint(Location location) {
        TrackPoint point = new TrackPoint(Calendar.getInstance().getTime(), location.getLatitude(), location.getLongitude(), location.getAccuracy());
        mPoints.add(point);
        return point;
    }

    public double getDistance()
    {
        double sum = 0.0;
        for (int i = 0; i < mPoints.size() - 1; i++) {
            TrackPoint p1 = mPoints.get(i);
            TrackPoint p2 = mPoints.get(i + 1);
            float[] result = new float[1];
            Location.distanceBetween(p1.getLatitude(), p1.getLongitude(), p2.getLatitude(), p2.getLongitude(), result);
            sum += result[0];
        }
        return sum;
    }

    public ArrayList<TrackPoint> getPoints() {
        return mPoints;
    }

    public String ToJSON() {
        StringWriter stringWriter = new StringWriter();
        JsonWriter jsonWriter = new JsonWriter(stringWriter);

        try {
            jsonWriter.beginObject();// begin root

            jsonWriter.name("Track").beginArray();
            for (TrackPoint point : mPoints) {
                jsonWriter.beginObject();
                jsonWriter.name("date").value(point.getTime().getTime());
                jsonWriter.name("latitude").value(point.getLatitude());
                jsonWriter.name("longitude").value(point.getLongitude());
                jsonWriter.name("accuracy").value(point.getLongitude());
                jsonWriter.endObject();
            }
            jsonWriter.endArray();// end websites

            // end root
            jsonWriter.endObject();

            return stringWriter.toString();

        } catch (IOException e) {
            e.printStackTrace();

            return null;
        }
    }

    public static Track FromJSON(String jsonString) {
        JSONObject jsonRoot = null;
        try {
            jsonRoot = new JSONObject(jsonString);

            JSONArray jsonArray = jsonRoot.getJSONArray("Track");
            Track track = new Track();

            for (int i = 0; i < jsonArray.length(); i++) {
                JSONObject pointJson = jsonArray.getJSONObject(i);
                track.mPoints.add(new TrackPoint(
                        new Date(pointJson.getLong("date")),
                        pointJson.getDouble("latitude"),
                        pointJson.getDouble("longitude"),
                        pointJson.getDouble("accuracy"))
                );
            }
            return track;
        } catch (JSONException e) {
            e.printStackTrace();
            return null;
        }
    }

    public static Track FromGPX(File file) {


        try {
            DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
            DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
            Document doc = dBuilder.parse(file);
            doc.getDocumentElement().normalize();
            Element gpx = (Element)doc.getElementsByTagName("gpx").item(0);
            NodeList trkList = gpx.getElementsByTagName("trk");
            Element trk = (Element)trkList.item(0);

            Track track = new Track();

            NodeList segments = trk.getElementsByTagName("trkseg");

            for (int temp = 0; temp < segments.getLength(); temp++) {
                Element segment = (Element) segments.item(temp);
                NodeList points = segment.getElementsByTagName("trkpt");
                for (int i = 0; i < points.getLength(); i++) {

                    Element point = (Element) points.item(i);

                    double latitude = Double.parseDouble(point.getAttribute("lat"));
                    double longitude = Double.parseDouble(point.getAttribute("lon"));

                    track.mPoints.add(new TrackPoint(Calendar.getInstance().getTime(), latitude, longitude, 0));
                }
            }

            return track;

        } catch (IOException e) {
            e.printStackTrace();
        } catch (ParserConfigurationException e)  {
        }catch (SAXException e) {
            e.printStackTrace();
        }

        return null;
    }

}

File: Android/app/src/main/java/com/max/spirihin/mytracksdb/Track.kt
package com.max.spirihin.mytracksdb

import android.location.Location
import android.util.JsonWriter
import android.util.Log
import org.json.JSONObject
import org.w3c.dom.Element
import java.io.File
import java.io.StringWriter
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.*
import javax.xml.parsers.DocumentBuilderFactory

class Track {

    //region properties
    val points: ArrayList<TrackPoint> = ArrayList()
    var id : Int = 0

    val distance: Double
        get() {
            //TODO we shouldn't compute this all the time, we need cache value and increment in addPoint
            var sum = 0.0
            for (i in 0 until points.size - 1) {
                val p1 = points[i]
                val p2 = points[i + 1]
                val result = FloatArray(1)
                Location.distanceBetween(p1.latitude, p1.longitude, p2.latitude, p2.longitude, result)
                sum += result[0]
            }
            return sum
        }

    /* full duration in seconds */
    val duration: Long
        get() = if (points.size < 2) 0 else (points.last().time.time - points.first().time.time) / 1000

    val timeStr: String
        get() {
            if (points.isEmpty())
                return ""
            val dateFormat: DateFormat = SimpleDateFormat("dd.MM - hh:mm:ss")
            return dateFormat.format(points[0].time)
        }
    //endregion

    //region public methods
    fun addPoint(location: Location): TrackPoint {
        val point = TrackPoint(Calendar.getInstance().time, location.latitude, location.longitude, location.accuracy.toDouble())
        points.add(point)
        return point
    }
    //endregion

    //region parsing
    fun toJSON(): String {
        val stringWriter = StringWriter()
        val jsonWriter = JsonWriter(stringWriter)
        jsonWriter.beginObject() // begin root
        jsonWriter.name("Track").beginArray()
        for (point in points) {
            jsonWriter.beginObject()
            jsonWriter.name("date").value(point.time.time)
            jsonWriter.name("latitude").value(point.latitude)
            jsonWriter.name("longitude").value(point.longitude)
            jsonWriter.name("accuracy").value(point.accuracy)
            jsonWriter.endObject()
        }
        jsonWriter.endArray()
        jsonWriter.endObject() // end root
        return stringWriter.toString()
    }

    companion object {
        fun fromGPX(file: File?): Track? {
            return try {
                val dbFactory = DocumentBuilderFactory.newInstance()
                val dBuilder = dbFactory.newDocumentBuilder()
                val doc = dBuilder.parse(file)
                doc.documentElement.normalize()
                val gpx = doc.getElementsByTagName("gpx").item(0) as Element
                val trkList = gpx.getElementsByTagName("trk")
                val trk = trkList.item(0) as Element
                val track = Track()
                val segments = trk.getElementsByTagName("trkseg")
                for (temp in 0 until segments.length) {
                    val segment = segments.item(temp) as Element
                    val points = segment.getElementsByTagName("trkpt")
                    for (i in 0 until points.length) {
                        val point = points.item(i) as Element
                        val latitude = point.getAttribute("lat").toDouble()
                        val longitude = point.getAttribute("lon").toDouble()
                        track.points.add(TrackPoint(Calendar.getInstance().time, latitude, longitude, 0.0))
                    }
                }
                track
            } catch (e: Exception) {
                Log.e("myLogs", e.message ?: "")
                null
            }
        }

        fun fromJSON(jsonString: String?): Track? {
            return try {
                val jsonRoot = JSONObject(jsonString ?: return null)
                val jsonArray = jsonRoot.getJSONArray("Track")
                val track = Track()
                for (i in 0 until jsonArray.length()) {
                    val pointJson = jsonArray.getJSONObject(i)
                    track.points.add(TrackPoint(
                            Date(pointJson.getLong("date")),
                            pointJson.getDouble("latitude"),
                            pointJson.getDouble("longitude"),
                            pointJson.getDouble("accuracy"))
                    )
                }
                track
            } catch (e: Exception) {
                Log.e("myLogs", e.message ?: "")
                null
            }
        }
    }
    //endregion
}